pkg/storage/chunk/cache/background.go

```diff
@@ -148,8 +148,13 @@ func (c *backgroundCache) Store(ctx context.Context, keys []string, bufs [][]byt
 		}
 
 		size := bgWrite.size()
-		newSize := c.size.Load() + int64(size)
+		// prospectively add new size
+		c.size.Add(int64(size))
+
+		newSize := c.size.Load()
 		if newSize > int64(c.sizeLimit) {
+			// subtract it since we've exceeded the limit
+			c.size.Add(-int64(size))
```

Comment:
`Sub` atomically subtracts from the wrapped int64 and returns the new value.           

```suggestion
			c.size.Sub(int64(size))
```